Added TWS parsing alongside some basic tests

SicklySilverMoon opened the pull request.

zyudneko commented on libchips/src/format-ccl.c:
```diff
@@ -78,11 +78,15 @@ static TileID const dat_tileid_map[] = {
+uint64_t read_uint64_le(uint8_t const* data) {
+  return (uint64_t) data[0] | ((uint64_t) data[1] << 8) | ((uint64_t) data[2] << 16) | ((uint64_t) data[3] << 24) |
+    ((uint64_t) data[4] << 32) | ((uint64_t) data[5] << 40) | ((uint64_t) data[6] << 48) | ((uint64_t) data[7] << 56);
 }
```
If this is now used in multiple files, the definition should be moved somewhere common too. Maybe `static inline` in formats.h or misc.h/misc.c, up to you

zyudneko commented on libchips/src/format-tws.c:
```diff
@@ -0,0 +1,297 @@
+}
+
+void TWSMetadata_free(TWSMetadata* self) {
+  if (self == NULL)
```
`NULL` checks aren't necessary for `free`

zyudneko commented on libchips/src/format-tws.c:
```diff
@@ -0,0 +1,297 @@
+    self->solutions = xrealloc(self->solutions, sizeof(TWSMetadata) * self->solutions_allocated);
+  }
+  self->solutions[self->solutions_n] = *level;
+  self->solutions_n++;
```
`+= 1`

zyudneko commented on libchips/src/format-tws.c:
```diff
@@ -0,0 +1,297 @@
+}
+
+void TWSSet_add_level(TWSSet* self, TWSMetadata* level) {
+  if (self->solutions_n + 1 > self->solutions_allocated) {
```
We probably should have a generic vector struct (like in `libnotcc`) instead of replicating vector logic everywhere we need it. Okay for now I suppose, I'll port over that code later

zyudneko commented on libchips/src/format-tws.c:
```diff
@@ -0,0 +1,297 @@
+  self->solutions_n++;
+}
+
+Result_TWSSetPtr get_error(TWSSet* self, const char* err_msg) {
```
`static`?

zyudneko commented on libchips/src/format-tws.c:
```diff
@@ -0,0 +1,297 @@
+int TWSMetadata_cmp(const void* lhs_v, const void* rhs_v) {
+  TWSMetadata const* lhs = lhs_v;
+  TWSMetadata const* rhs = rhs_v;
+  if (lhs->level_num < rhs->level_num) {
+    return -1;
+  } else if (lhs->level_num > rhs->level_num) {
+    return 1;
+  } else {
+    return 0;
+  }
```
Are there overflow issues with just `return lsh->level_num - rhs->level_num`?

SicklySilverMoon replied:
Potentially yes, as I believe the TWS format allows the full `uint16_t` range, as well as unsigned subtraction to signed

zyudneko replied:
Okay, I suppose it can be left as-is

zyudneko commented on libchips/src/format-tws.c:
```diff
@@ -0,0 +1,297 @@
+  if (self->set_name != NULL)
+    free(self->set_name);
+  if (self->solutions) {
+    for (uint32_t i = 0; i < self->solutions_n; i++) {
```
`+= 1`

zyudneko commented on libchips/src/format-tws.c:
```diff
@@ -0,0 +1,297 @@
+}
+
+TWSMetadata const* TWSSet_get_level_solution(TWSSet const* self, uint16_t level_num) {
+  for (uint16_t i = 0; i < self->solutions_n; i++) {
```
`+= 1`

zyudneko commented on libchips/src/format-tws.c:
```diff
@@ -0,0 +1,297 @@
+  }
+}
+
+  //https://www.muppetlabs.com/~breadbox/software/tworld/tworldff.html#3
```
Remove leading spaces, add space after `//` (applies to all comments)

zyudneko commented on libchips/src/format-tws.c:
```diff
@@ -0,0 +1,297 @@
+      if (size <= 16)
+        return get_error(set, "Not enough data for set name string.");
+      data += 10;
+      size -= 16;
```
Should probably advance `data` once

SicklySilverMoon replied:
Going to request elaboration on what you specifically mean by this one please

zyudneko replied:
`data += 16;` instead of the split `+= 10` and `+= 6`

zyudneko commented on libchips/src/format-tws.c:
```diff
@@ -0,0 +1,297 @@
+    while (size == 0) {
+      size = read_uint32_le(data);
+      data += 4;
+    }
```
Move `assert_data_avail` inside the loop

zyudneko commented on libchips/src/format-tws.c:
```diff
@@ -0,0 +1,297 @@
+        data += 4;
+        size -= 10;
+
+        assert(((GameInput) DIRECTION_NIL) == 0);
```
`static_assert`?

zyudneko commented on libchips/test/ccl/format-ccl.cpp:
```diff
@@ -0,0 +1,346 @@
+#include "formats.h"
+}
+
+char const* titles[] = {
```
Should only have one array of string + time pair structs

zyudneko commented on libchips/test/ccl/format-ccl.cpp:
```diff
@@ -0,0 +1,346 @@
+  600,
+};
+
+const uint8_t CCLP1_ccl[] = {
```
Don't we already have this in `ccl_embeds.h`?

zyudneko commented on libchips/test/ccl/format-ccl.cpp:
```diff
@@ -0,0 +1,346 @@
+
+    LevelSet* set = res.value;
+    EXPECT_EQ(set->levels_n, 149);
+    EXPECT_EQ(set->levels_n, 149);
```
Duplicate line?

zyudneko commented on libchips/test/data/ccl/ccl_embeds.h:
```diff
@@ -0,0 +1,29 @@
+#embed "CCLP5.ccl"
+};
+
+#endif //LIBCHIPS_CCL_EMBEDS_H
```
EOF newline pls

zyudneko commented on libchips/test/logic/cclp1_tws_playthrough.cpp:
```diff
@@ -0,0 +1,62 @@
+
+namespace {
+  TEST(CCLP1TWS, LoadAndPlayMS) {
+    Result_LevelSetPtr res = parse_ccl(CCLP1_ccl, sizeof(CCLP1_ccl));
```
1. In the future I'd like to have a CLI for running any CCL with any TWS (like I have with `libnotcc` and its CLI), but for now this is fine to have.
2. Please split the boilerplate into a function. You're missing some `*_free`s in the Lynx code, but the generic function will fix that.

zyudneko commented on libchips/test/logic/cclp1_tws_playthrough.cpp:
```diff
@@ -0,0 +1,62 @@
+      TWSMetadata* solution = &tws->solutions[i];
+      for (size_t j = 0; j < solution->num_ticks; j++) {
+        level->game_input = solution->inputs[j];
+        ms_logic.tick_level(level);
```
Use `Level_tick`, it has some extra logic

zyudneko commented on libchips/test/tws/format-tws.cpp:
```diff
@@ -0,0 +1,466 @@
+    TWSSet_free(set);
+  }
+
+}
```
EOL newline pls

zyudneko commented on libchips/test/CMakeLists.txt:
```diff
@@ -0,0 +1,27 @@
+set(gtest_force_shared_crt ON CACHE BOOL "" FORCE)
+FetchContent_MakeAvailable(googletest)
+
+include_directories(../src) #allow referencing source as if we were in the src dir
```
Okay, but note for me - we should split (public) header files into separate `include` dir

zyudneko commented on libchips/test/CMakeLists.txt:
```diff
@@ -0,0 +1,27 @@
+include_directories(../src) #allow referencing source as if we were in the src dir
+include_directories(.)
+
+file(GLOB_RECURSE GLOB_SRCS ../src/*.h ../src/*.c ../src/*.cpp)
```
We should link with `libchips` library, not build with its source

TheGLander reviewed (changes requested):
Code's mostly fine, thanks for finding the time to work on this!

SicklySilverMoon commented:
Requesting review again, although at present loading a TWS file preloads all possible moves, I've made changes to many places such as logic (as I got ahead of myself) and would greatly prefer to not have to start reconciling in the event of further updates to the main branch

zyudneko commented on libchips/src/format-tws.c:
```diff
@@ -205,93 +288,84 @@ Result_TWSSetPtr parse_tws(uint8_t const* data, size_t data_len) {
-          }
-        }
+        assert_data_avail(size);
+        level.compressed_inputs = xmalloc(sizeof(CompressedInputList));
```
I don't think we should be dynamically allocating this. I understand this is probably to make it clearly nullable, but I think we should represent that differently somehow (maybe a Rust-like `Opt` union, or a null `bytes` pointer representing no value?)

zyudneko commented on libchips/src/format-tws.c:
```diff
@@ -205,93 +288,84 @@ Result_TWSSetPtr parse_tws(uint8_t const* data, size_t data_len) {
+  if (new_size <= self->count) {
+    self->allocated = new_size;
+    self->count = new_size;
+    self->inputs = xrealloc(self->inputs, sizeof(GameInput) * self->allocated);
```
Should just set `count` without reallocating, user can use `shrink` when they know they aren't gonna be appending any more

zyudneko commented on libchips/src/format-tws.c:
```diff
@@ -205,93 +288,84 @@ Result_TWSSetPtr parse_tws(uint8_t const* data, size_t data_len) {
+    self->inputs = xrealloc(self->inputs, sizeof(GameInput) * self->allocated);
+  } else {
+    self->allocated = new_size;
+    self->inputs = xrealloc(self->inputs, sizeof(GameInput) * self->allocated);
```
Should follow same reallocation logic as resizing in `append` (which should be split to a separate helper function)

zyudneko commented on libchips/src/format-tws.c:
```diff
@@ -205,93 +288,84 @@ Result_TWSSetPtr parse_tws(uint8_t const* data, size_t data_len) {
+    self->inputs = xrealloc(self->inputs, sizeof(GameInput) * self->allocated);
+    for (size_t i = self->count; i < self->allocated; i += 1) {
+      self->inputs[i] = DIRECTION_NIL;
+    }
```
`static_assert(DIRECTION_NIL == 0)` + `memset`?

SicklySilverMoon committed:
Added TWS parsing and an interface/API for it.
Added automated testing using the public TWS and CCLPs.
Corrected some MS logic issues.

SicklySilverMoon commented:
Now that commits have been squashed into one using the above (see full original history at [https://github.com/SicklySilverMoon/tworld3/tree/pr-tws](https://github.com/SicklySilverMoon/tworld3/tree/pr-tws)), it is being merged.
